File: Leetcode/cpp/Merge Intervals.cc
const int INF = 1 << 25;

struct cmp {
    Interval it;
    cmp(const Interval& i_it)
    {
        it = i_it;
    }
    bool operator ()(const Interval& a, const Interval& b)
    {
        return a.start < b.start;
    }
};

 
class Solution {
public:
    vector<Interval> merge(vector<Interval> &intervals)
    {
        vector<Interval> res;
        if (!intervals.size()) {
            return res;
        } else {
            intervals.push_back(Interval(INF, INF));
        }
        sort(intervals.begin(), intervals.end(), cmp(intervals[0]));
        int st  = intervals[0].start;
        int end = intervals[0].end;
        for (int i = 0; i < intervals.size(); i++) {
            if (in_interval(intervals[i], st, end)) {
                end = max(end, intervals[i].end);
            } else {
                res.push_back(Interval(st, end));
                st  = intervals[i].start;
                end = intervals[i].end; 
            }
        }
        return res;
    }

    bool in_interval(const Interval& interval, int st, int end)
    {
        if (1LL * (interval.start - st) * (interval.start - end) <= 0) {
            return true;
        } else {
            return false;
        }
    }
};

File: Leetcode/python/Merge Intervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    # @param intervals, a list of Interval
    # @return a list of Interval
    def merge(self, intervals):
        INF = 0xdeadbeef
        intervals.append(Interval(INF, INF))
        intervals.sort(key=lambda x: x.start)
        res = []
        cur = intervals[0]
        
        for interval in intervals:
            if interval.start > cur.end:
                res.append(cur)
                cur = interval
            else:
                cur.end = max(interval.end, cur.end)
        
        return res
            
